src/core/core.datasetController.js

```diff
@@ -842,72 +853,147 @@ helpers.extend(DatasetController.prototype, {
 	/**
 	 * @private
 	 */
-	_resolveDataElementOptions: function(index) {
+	_resolveDataElementOptions: function(index, mode) {
 		const me = this;
+		const active = mode === 'active';
 		const cached = me._cachedDataOpts;
-		if (cached) {
-			return cached;
+		if (cached[mode]) {
+			return cached[mode];
 		}
 		const chart = me.chart;
 		const datasetOpts = me._config;
 		const options = chart.options.elements[me.dataElementType.prototype._type] || {};
 		const elementOptions = me._dataElementOptions;
 		const values = {};
-		const context = {
-			chart: chart,
-			dataIndex: index,
-			dataset: me.getDataset(),
-			datasetIndex: me.index
-		};
-		const info = {cacheable: true};
-		let keys, i, ilen, key;
+		const context = me._getContext(index, active);
+		const info = {cacheable: !active};
+		let keys, i, ilen, key, value, readKey;
 
 		if (helpers.isArray(elementOptions)) {
 			for (i = 0, ilen = elementOptions.length; i < ilen; ++i) {
 				key = elementOptions[i];
-				values[key] = resolve([
-					datasetOpts[key],
-					options[key]
+				readKey = active ? 'hover' + key.charAt(0).toUpperCase() + key.slice(1) : key;
+				value = resolve([
+					datasetOpts[readKey],
+					options[readKey]
 				], context, index, info);
+				if (value !== undefined) {
+					values[key] = value;
+				}
 			}
 		} else {
 			keys = Object.keys(elementOptions);
 			for (i = 0, ilen = keys.length; i < ilen; ++i) {
 				key = keys[i];
-				values[key] = resolve([
-					datasetOpts[elementOptions[key]],
-					datasetOpts[key],
-					options[key]
+				readKey = active ? 'hover' + key.charAt(0).toUpperCase() + key.slice(1) : key;
+				value = resolve([
+					datasetOpts[elementOptions[readKey]],
+					datasetOpts[readKey],
+					options[readKey]
 				], context, index, info);
+				if (value !== undefined) {
+					values[key] = value;
+				}
 			}
 		}
 
 		if (info.cacheable) {
-			me._cachedDataOpts = Object.freeze(values);
+			values.$shared = true;
+			cached[mode] = values;
```

Comment:
it's not clear to me why the options must be cached by mode. there's a different set of options for each mode? can you add a code comment explaining this part?

Reply:
Done